providers/gpgpu/bin/gpu_passthrough.py

```diff
@@ -0,0 +1,576 @@
+        proc = subprocess.run(
+            shlex.split(cmd),
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+            stdin=subprocess.DEVNULL,
+            universal_newlines=True,
+            check=False,
+        )
+
+        logging.debug("Command: %s", cmd)
+        if proc.stdout:
+            logging.debug(" STDOUT: %s", proc.stdout)
+        if log_stderr and proc.stderr:
+            logging.debug(" STDERR: %s", proc.stderr)
+
+        if check:
+            proc.check_returncode()
+
+        return proc
+
+    def download_image(self, url, filename):
+        """Downloads LXD files for same release as host machine."""
+        logging.debug("Attempting download of %s from %s", filename, url)
+        try:
+            urllib.request.urlretrieve(url, filename)
+        except (
+            IOError,
+            OSError,
+            urllib.error.HTTPError,
+            urllib.error.URLError,
+        ) as e:
+            logging.exception(e)
+        except ValueError as verr:
+            logging.exception(verr)
+
+        if not os.path.isfile(filename):
+            raise FileNotFoundError(filename)
+
+    def insert_images(self):
+        """Insert LXD template and image images."""
+        if self.template and self.image:
+            logging.debug("Importing images into LXD")
+            self.run(
+                "lxc image import {} rootfs {} --alias {}".format(
+                    self.template, self.image, self.image_alias.hex
+                ),
+                check=True,
+            )
+        else:
+            logging.debug("No local images, attempting import from remote")
+            retries = 2
+            for _ in range(retries):
+                proc = self.run(
+                    "lxc image copy {}{} local: --alias {}".format(
+                        self.remote, RELEASE, self.image_alias.hex
+                    )
+                )
+                if proc.returncode == 0:
+                    return
+                logging.error("Error encountered while importing images")
+                logging.error("Attempting up to %d times", retries)
+            raise RuntimeError("Images could not be inserted")
+
+    def init_lxd(self):
+        """Initializes LXD."""
+        if self.run("lxd waitready --timeout=5").returncode != 0:
+            self.run("lxd init --auto", check=True)
+        self.insert_images()
+
+    def cleanup(self):
+        """Cleans up instance."""
+        self.run("lxc image delete {}".format(self.image_alias.hex))
+        self.run("lxc delete --force {}".format(self.name))
+
+    def launch(self, options: Optional[List] = None):
+        """Sets up and creates the instance."""
+        if not options:
+            options = []
+
+        cmd = "lxc launch {} {}".format(self.image_alias.hex, self.name)
+        if options:
+            cmd = "{} {}".format(cmd, " ".join(options))
+        self.run(cmd, check=True)
```

Comment:
```suggestion
    def launch(self, options: Optional[List] = []):
        """Sets up and creates the instance."""
        cmd = ["lxc", "launch", self.image_alias.hex, self.name]
        cmd += options
        self.run(shlex.join(cmd))
```

Reply:
Resolved, but we should not be using mutable default arguments.

Reply:
You should not mutate default arguments in my opinion, I don't agree one should completely abandon the thought of default values if they are not immutable and implement them manually via a default value + if else.